src/symmetric_crypto.cpp

```diff
@@ -0,0 +1,349 @@
+                break;
+            default:
+                throw MoCOCrWException("Not yet implemented cipher mode.");
+        }
+
+        _EVP_CipherInit_ex(_ctx.get(),
+                                    EVPCipherConstructor(),
+                                    nullptr,
+                                    nullptr,
+                                    nullptr,
+                                    this->_operation == Operation::Encryption);
+
+        // Check id key size matches cipher block size
+        size_t expectedKeySize = _EVP_CIPHER_CTX_key_length(_ctx.get());
+        if (secretKey.size() != expectedKeySize) {
+            auto formatter = boost::format("Invalid size of Key %d bytes. Must be %d bytes.");
+            formatter % iv.size() % expectedKeySize;
+            throw MoCOCrWException(formatter.str());
+        }
+
+        // Check IV length and adjust if cipher supports it
+        switch (mode) {
+            case SymmetricCipherMode::GCM: {
+                _EVP_CIPHER_CTX_ctrl(_ctx.get(), EVP_CTRL_GCM_SET_IVLEN, _iv.size(),
+                                              nullptr);
+            } break;
+            case SymmetricCipherMode::CTR:
+                //[[fallthrough]];
+            case SymmetricCipherMode::CBC: {
+                size_t expectedIVSize = _EVP_CIPHER_CTX_iv_length(_ctx.get());
+                if (_iv.size() != expectedIVSize) {
+                    auto formatter = boost::format(
+                            "Invalid size of IV %d bytes. Must be %d bytes.");
+                    formatter % _iv.size() % expectedIVSize;
+                    throw MoCOCrWException(formatter.str());
+                }
+            } break;
+        }
+
+        _EVP_CIPHER_CTX_reset(_ctx.get());
+
+        _EVP_CipherInit_ex(_ctx.get(),
+                                    EVPCipherConstructor(),
+                                    nullptr,
+                                    secretKey.data(),
+                                    iv.data(),
+                                    this->_operation == Operation::Encryption);
+
+        switch (padding) {
+            case SymmetricCipherPadding::PKCS:
+                // OpenSSL uses PKCS padding by default.
+                break;
+            case SymmetricCipherPadding::NO:
+                _EVP_CIPHER_CTX_set_padding(_ctx.get(), 0);
+                break;
+        }
+    };
+
+    void update(const std::vector<uint8_t> &message)
+    {
+        if (_isFinished) {
+            throw MoCOCrWException(
+                    "Further calls to update() are not allowed once finish() was called.");
+        }
+
+        if (message.size() > (std::numeric_limits<int>::max() - EVP_MAX_BLOCK_LENGTH)) {
+            throw MoCOCrWException("Message is too big.");
+        }
+        int processingChunkSize = message.size();
+        std::vector<uint8_t> processedChunk(processingChunkSize + EVP_MAX_BLOCK_LENGTH);
```

Comment:
Likewise we should probably cast `processingChunkSize` to `std::size_t` before the addition to prevent a potential wraparound and sign-extension before the implicit upcast.

Reply:
see comment above